client/my-sites/themes/themes-selection.jsx

```diff
@@ -196,13 +199,20 @@ const ConnectedThemesSelection = connect(
 			number
 		};
 
+		const lastQuery = getLastThemeQuery( state, sourceSiteId ) || query;
+		const themesCount = getThemesFoundForQuery( state, sourceSiteId, lastQuery );
+
+		if ( ! isEqual( omit( query, PAGINATION_QUERY_KEYS ), omit( lastQuery, PAGINATION_QUERY_KEYS ) ) ) {
+			query.page = 1;
+		}
+
 		return {
 			query,
 			source: sourceSiteId,
 			siteSlug: getSiteSlug( state, siteId ),
-			themes: getThemesForQueryIgnoringPage( state, sourceSiteId, query ) || [],
-			themesCount: getThemesFoundForQuery( state, sourceSiteId, query ),
-			isRequesting: isRequestingThemesForQuery( state, sourceSiteId, query ),
+			themes: getThemesForQueryIgnoringPage( state, sourceSiteId, lastQuery ) || noThemes,
```

Comment:
This is a very good question and it has very interesting answer.

react-redux adds for each component `shouldComponentUpdate` method that checks if state or props have changed in a way that requires re-rendering the component. This function works by doing a shallow check - it only checks references. Empty array literal `[]` is equal to `new Array` and that means that every time we are creating a new empty array that has different reference than the previous empty array. This means that from the perspective of the`shouldComponentUpdate` we need to re-render when in fact nothing has changed. `noThemes` on the other hand has always the same reference.